prow/apis/prowjobs/v1/types.go

```diff
@@ -368,9 +368,72 @@ type DecorationConfig struct {
 	// which is going to be used for fetching a private repository.
 	OauthTokenSecret *OauthTokenSecret `json:"oauth_token_secret,omitempty"`
 
-	// CensorSecrets determines if the sidecar should be instructed to censor
-	// secret data that is otherwise mounted to the ProwJob Pod
+	// CensorSecrets is deprecated, set CensoringOptions instead
 	CensorSecrets *bool `json:"censor_secrets,omitempty"`
+
+	// CensoringOptions exposes options for censoring output logs and artifacts.
+	// The presence of this field will enable censoring for a job.
+	CensoringOptions *CensoringOptions `json:"censoring_options,omitempty"`
+}
+
+type CensoringOptions struct {
+	// CensoringConcurrency is the maximum number of goroutines that should be censoring
+	// artifacts and logs at any time. If unset, defaults to 10.
+	CensoringConcurrency *int64 `json:"censoring_concurrency,omitempty"`
+	// CensoringBufferSize is the size in bytes of the buffer allocated for every file
+	// being censored. We want to keep as little of the file in memory as possible in
+	// order for censoring to be reasonably performant in space. However, to guarantee
+	// that we censor every instance of every secret, our buffer size must be at least
+	// two times larger than the largest secret we are about to censor. While that size
+	// is the smallest possible buffer we could use, if the secrets being censored are
+	// small, censoring will not be performant as the number of I/O actions per file
+	// would increase. If unset, defaults to 10MiB.
+	CensoringBufferSize *int `json:"censoring_buffer_size,omitempty"`
+
+	// IncludeDirectories are directories which should have their content censored. If
+	// present, only content in these directories will be censored. Entries in this list
+	// are parsed with the go-zglob library, allowing for globbed matches.
+	IncludeDirectories []string `json:"include_directories,omitempty"`
+
+	// ExcludeDirectories are directories which should not have their content censored. If
+	// present, content in these directories will not be censored even if the directory also
+	// matches a glob in IncludeDirectories. Entries in this list are parsed with the go-
+	// zglob library, allowing for globbed matches.
+	ExcludeDirectories []string `json:"exclude_directories,omitempty"`
```

Comment:
Coming at this with very little context... as in, if I'm just looking at these options, I don't understand this feature:
- what path are these directories relative to?
- what's getting censored when/where again? this is artifacts and logs right? happening concurrently as the test is running? or before artifacts are uploaded?
- is it per container? per job?
- is this on the node running the job? is this only supported by `agent: kubernetes` jobs?

Reply:
> what path are these directories relative to?

The only path that users ever do anything relative to, the `$ARTIFACTS` path - but I can clarify that in the GoDoc. I honestly don't think it's worth trying to include or exclude files from this, I imagine the compute/time taken to load & censor even large build files is tiny and the mental overhead of figuring out what might and might not contain sensitive data and to what the paths are relative to, etc, outweigh the costs ... but the feature was requested, so here is is :)

> what's getting censored when/where again?

This is a two-fold answer. First, I imagine it's best if this were opaque to the user. Then, they can just use the thing and forget about it. Second, it depends on how well the system can auto-detect this for people. Right now, the impl will censor any data present in k8s secrets mounted to the containers in the Pod that execute user code.

> this is artifacts and logs right? happening concurrently as the test is running? or before artifacts are uploaded?

This is an option of `sidecar`, which handles artifacts and logs pushed to GCS, so I think in context you kind of know the answers to all of these if you're actually configuring it. Censoring happens before it's uploaded, but from a user's perspective a Prow Job has no artifacts until they are uploaded so that question is kind of moot.

> is it per container? per job?

All Prow configuration is per job. No ability to configure options per container exists for any part of the system. IIRC only one `sidecar` runs even when more than one container exists in the job anyway, so there's no real clear answer to this question. Again, I'd wager the user does not care about this distinction.

> is this on the node running the job?

This is part of Prow k8s job decoration, so yes like with any other part of the larger system it runs on Kubernetes jobs only.

Reply:
Please clarify godocs to indicate that paths are relative to `$ARTIFACTS`.